File: src/main/java/pl/training/workshop/m8/s02_stranglerfig/step2/LegacyCinema.java
package pl.training.workshop.m8.s02_stranglerfig.step2;

import java.math.BigDecimal;
import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;

import pl.training.workshop.m8.s02_stranglerfig.BookingLedger;
import pl.training.workshop.m8.s02_stranglerfig.CinemaApi;
import pl.training.workshop.shared.Money;

/**
 * Krok 2 (bez zmian): stary system. Od kroku 1 klienci widzą go tylko przez CinemaFacade.
 */
public final class LegacyCinema implements CinemaApi {
    private final BookingLedger ledger;

    public LegacyCinema(BookingLedger ledger) {
        this.ledger = ledger;
    }

    @Override
    public String book(String email, String title, int format, int tickets, boolean web) {
        if (tickets <= 0) {
            return "ERROR: no seats";
        }
        double p = format == 1 ? 25.00 : format == 2 ? 32.00 : 40.00;
        double sum = p * tickets;
        if (tickets >= 10) {
            sum = sum - sum * 0.10;
        }
        double fees = web ? 2.00 * tickets : 0;
        String id = ledger.nextId();
        ledger.add(new BookingLedger.Booking(id, email, title, tickets,
                new Money(BigDecimal.valueOf(sum)), new Money(BigDecimal.valueOf(fees))));
        return id;
    }

    @Override
    public String report() {
        Map<String, double[]> byTitle = new TreeMap<>();
        double fees = 0;
        int count = 0;
        for (BookingLedger.Booking b : ledger.all()) {
            double[] row = byTitle.computeIfAbsent(b.title(), k -> new double[2]);
            row[0] = row[0] + b.tickets();
            row[1] = row[1] + b.ticketsValue().amount().doubleValue();
            fees = fees + b.fees().amount().doubleValue();
            count = count + b.tickets();
        }
        StringBuilder sb = new StringBuilder("RAPORT\n");
        double revenue = 0;
        for (Map.Entry<String, double[]> e : byTitle.entrySet()) {
            sb.append(e.getKey()).append(": ").append((int) e.getValue()[0]).append(" bil., ")
                    .append(String.format(Locale.ROOT, "%.2f", e.getValue()[1])).append("\n");
            revenue = revenue + e.getValue()[1];
        }
        sb.append("Biletow: ").append(count).append("\n");
        sb.append("Przychod z biletow: ").append(String.format(Locale.ROOT, "%.2f", revenue)).append("\n");
        sb.append("Oplaty rezerwacyjne: ").append(String.format(Locale.ROOT, "%.2f", fees)).append("\n");
        return sb.toString();
    }
}

File: typescript/src/workshop/m8/s02_stranglerfig/step3/LegacyCinema.ts
import { Decimal } from 'decimal.js';

import { Money } from '../../../shared/Money.js';
import { Booking, type BookingLedger } from '../BookingLedger.js';
import type { CinemaApi } from '../CinemaApi.js';

/**
 * Krok 3 (bez zmian): stary system. Od kroku 1 klienci widzą go tylko przez CinemaFacade.
 */
export class LegacyCinema implements CinemaApi {
  constructor(private readonly ledger: BookingLedger) {}

  book(email: string, title: string, format: number, tickets: number, web: boolean): string {
    if (tickets <= 0) {
      return 'ERROR: no seats';
    }
    const p = format === 1 ? 25.00 : format === 2 ? 32.00 : 40.00;
    let sum = p * tickets;
    if (tickets >= 10) {
      sum = sum - sum * 0.10;
    }
    const fees = web ? 2.00 * tickets : 0;
    const id = this.ledger.nextId();
    this.ledger.add(new Booking(id, email, title, tickets,
      new Money(new Decimal(sum)), new Money(new Decimal(fees))));
    return id;
  }

  report(): string {
    const byTitle = new Map<string, number[]>();
    let fees = 0;
    let count = 0;
    for (const b of this.ledger.all()) {
      let row = byTitle.get(b.title);
      if (row === undefined) {
        row = [0, 0];
        byTitle.set(b.title, row);
      }
      row[0] = row[0]! + b.tickets;
      row[1] = row[1]! + b.ticketsValue.amount.toNumber();
      fees = fees + b.fees.amount.toNumber();
      count = count + b.tickets;
    }
    let sb = 'RAPORT\n';
    let revenue = 0;
    for (const title of [...byTitle.keys()].sort()) {
      const row = byTitle.get(title)!;
      sb += title + ': ' + Math.trunc(row[0]!) + ' bil., ' + row[1]!.toFixed(2) + '\n';
      revenue = revenue + row[1]!;
    }
    sb += 'Biletow: ' + count + '\n';
    sb += 'Przychod z biletow: ' + revenue.toFixed(2) + '\n';
    sb += 'Oplaty rezerwacyjne: ' + fees.toFixed(2) + '\n';
    return sb;
  }
}
